zerver/lib/response.py

```diff
@@ -1,11 +1,76 @@
-from typing import Any, List, Mapping, Optional
+from typing import Any, Dict, Iterator, List, Mapping, Optional
 
 import orjson
 from django.http import HttpRequest, HttpResponse, HttpResponseNotAllowed
 
 from zerver.lib.exceptions import JsonableError, UnauthorizedError
 
 
+class MutableJsonResponse(HttpResponse):
+    def __init__(
+        self,
+        data: Dict[str, Any],
+        *,
+        content_type: str,
+        status: int,
+    ) -> None:
+        # Mirror the behavior of Django's TemplateResponse and pass an
+        # empty string for the initial content value. Because that will
+        # set _needs_serialization to False, we initialize it to True
+        # after the call to super __init__.
+        super().__init__("", content_type=content_type, status=status)
+        self._data = data
+        self._needs_serialization = True
+
+    def get_data(self) -> Dict[str, Any]:
+        """Get data for this MutableJsonResponse. Calling this method
+        after the response's content has already been serialized
+        will mean the next time the response's content is accessed
+        it will be reserialized."""
```

Comment:
Might be worth adding a few words explaining why we do this -- I think the idea is that the caller might mutate the data?